### DISCUSSION TITLE
Reduce operation

### DISCUSSION THREAD

**Discussion by innovate-invent** [jmespath-community/jmespath.spec#48] [Ideas]
We have map to apply the same operation resulting in the same size output, but we don't have a reduce operation that allows aggregating an arbitrary dataset. 

The reduce operation is a bit complicated as the expression that is being evaluated for each input must accept two inputs, the aggregate so far and the current input. There is currently no way in JMESPath to refer to more than one item in context `@`. The `let()` function is slated to resolve that for most use cases, but it can not be used recursively.

To contrast `map(&expression, array)`, expression accesses the current item via `@` when being evaluated. 

We could do some trickery with an implied `let()` around each evaluation of the expression with a reserved identifier for the aggregate, but that is a bit cludgey. 

We could equally introduce an additional symbol rather than an identifier, but that is not much better and it adds single use grammar for a function. 

The expression could have a wrapped context where `@` is an object like `{"acc":.., "cur":..}` and the expression must explicitly refer to each, but that is not much better than the first suggestion, putting the current item behind an additional identifier as well. At least with this there is no possibility of the aggregate identifier colliding with an identifier in the current item.

I am interested to hear alternatives or if any of these suggestions are preferred.

**Comment by innovate-invent** [comment]
[My post](https://github.com/jmespath-community/jmespath.jep/discussions/47#discussioncomment-2385057) in the other discussion goes into more detail but I think we should consider implementing this as syntax.

As far as I can tell `map(&expression, a)` is redundant to `a[*].expression`. We could potentially extend the syntax to something like `a[&].merge(acc, cur)` for a reducing projection. I also think the wrapped context is the best way to move forward, it is a bit odd to suddenly have identifiers in a query that are not declared in the query or data though.

**Reply by innovate-invent** [reply]
I missed a rather important detail, we need some way to express the initial value of acc. 
Possibly stuff the initial value in the projection operator as `a[&{}].merge(acc, cur)`.

**Reply by innovate-invent** [reply]
I am struggling to determine what `@` should be within the context of the initial value expression. 
it could be the value of `a`, but I am not sure why we would want to access the object we are about to iterate over, other than possibly evaluating the length of `a` during init.
Having it be the first value of `a` is redundant, because that will be available during iteration.  
What would be a generally useful value of `@` while initializing a reduce operation over an object?

**Reply by innovate-invent** [reply]
Actually, the value of `a` would be ideal as there could be conditional init logic based on the input value.

**Reply by springcomp** [reply]
I think this is consistent with the current implementation. `@` is already the value of `a` in a projection operation.

|Given|Expression|Result|Comment|
|---|---|---|---|
| `[{"foo": "bar"}, {"baz": "qux"}]` | `[@]` | `[ [{"foo": "bar"}, {"baz": "qux"}] ] ` | The current value is the array itself.
| `[{"foo": "bar"}, {"baz": "qux"}]` | `@[@]` | `[ [{"foo": "bar"}, {"baz": "qux"}] ]` | This gives the same result as above.
| `[{"foo": "bar"}, {"baz": "qux"}]` | `@[@][0]` | `[{"foo": "bar"}, {"baz": "qux"}]` | This gives back the original object.

**Reply by innovate-invent** [reply]
I am increasingly unhappy with the magic `{acc: , cur: }` context within the reduce projection, it feels cludgy. What do you think of co-opting the `$` root reference proposal, where by we define `$` depending on its context. In the root context it can refer to the root, but within reducing projections it could refer to `acc`, where `@` refers to `cur`. 

If we were to go in that direction, I would want other definitions of `$` so it is not just those uses. Perhaps in mapping projections it could refer to the key or index of the current value. What do you think?

**Reply by innovate-invent** [reply]
I think given https://github.com/jmespath-community/jmespath.spec/discussions/68 we should use a different symbol for the reduce operation. Perhaps `[# expression ]`?

**Reply by springcomp** [reply]
I‘m not really statisfied with overloading a symbol with different behaviour.
Maybe that’s not a valid concern, but maybe we would also want to support the root reference within a reduce projection...

I agree that the magic `{acc: , cur: }` context within the reduce projection feels like a _hack_. For instance, currently we support:

```
search(sum([].age), [{"name": "Mark", "age": 50}, {"name": "Ted", "age": 42}]) -> 92
```

Incidentally, we would still not be able to express an hypothetical `reduce` function:

```
search(reduce([].age, 0, &acc + cur), [{"name": "Mark", "age": 50}, {"name": "Ted", "age": 42}]) -> 92
```

So what we are really after is the concept of _lambda_ that `expression-type` falls short of. That is, we need a means to _name_  arguments that would be specified to the expression. We could have something like this added to the grammar:

```
lambda-expression = lambda-arguments "=>" expression-type
lambda-arguments = "(" identifier *( "," identifier ) ")" 
```

Which would allow something like:

```
reduce([].age, 0, (acc, cur) => &acc + cur)
```

Where the order and semantics of the arguments would be defined by the specification of the  `reduce` function.
If we now want to express this with a `reduce-projection` instead of a function, we still have to specify the symbol we want to use.

An `expression-type` expression would be a simplified syntax for a `lambda-expression` that does not take any arguments at all. That is, something like:

```
() => &expr
```

**Reply by springcomp** [reply]
I thinks any symbol that is not already used would be all right within `reduce-projection` expressions.

However, as a big fan of PowerShell, I like the following symbols `?` and `%` used respectively for the `where` and `foreach` filters. As JMESPath already defines `?` for the `where` filter, I like the symmetry it would provide if we think of a `reduce-projection` as being somewhat akin to a `foreach` filter (although it strictly is more akin to `map`, really).

That would allow us to make the following two expressions equivalent:

```
search(sum([].age), [{"name": "Mark", "age": 50}, {"name": "Ted", "age": 42}]) -> 92
search([%0].age.(acc, cur) => &acc + cur), [{"name": "Mark", "age": 50}, {"name": "Ted", "age": 42}]) -> 92
```
Although we touch on the subject of readibility and ease of use.

On another note, if we accept `lambda-expression` in the grammar, we would perhaps need a way to express discarded arguments.
In languages that support this concept, `_` is a popular symbol but  `_` is a valid identifier in JMESPath. So we would need to find an alternative.

**Reply by innovate-invent** [reply]
It isn't clear why this wouldn't work:
```
search(reduce([].age, 0, &acc + cur), [{"name": "Mark", "age": 50}, {"name": "Ted", "age": 42}]) -> 92
```

Can you elaborate more?

**Reply by springcomp** [reply]
It definitely would.

However, it also has this "magic" use of `acc` and `cur` identifiers that pop out of nowhere. The magic is there irrespective of the fact reduce is implemented as a function or as syntax.

That's what I addressed specifically with the concept of lambda.

**Reply by innovate-invent** [reply]
I see, this is why I was [proposing using the `$` symbol](https://github.com/jmespath-community/jmespath.spec/discussions/18#discussioncomment-2708078) to represent `acc`. I think the `$` symbol should be a context sensitive reference that refers to a "greater" context. This is the root node by default, the key value in a projection (if you think of the key as the parent of the value), and the accumulated value in a reduce projection. We could also potentially use it to refer to the defining context of an `expression-type`.

This removes the need for the magic identifiers with a more consistently defined symbol.

I am not thrilled about the idea of introducing a lambda syntax. I appreciate you are trying for a more declarative syntax to remove the "magic" identifiers, but I can't see a use case beyond a single additional special identifier other than `@`. This would also heavily overlap with my [reusable expressions](https://github.com/jmespath-community/jmespath.spec/discussions/68) proposal.

**Comment by innovate-invent** [comment]
To summarize a few different discussions I think the best standing proposal is for the reduce operation to be:
```
reducing-projection = "[" "%" [expression] "]"
```

Like the projection operation, the expression following the reducing projection is applied to every value of an array or map. The `$` symbol represents the accumulated/previous value of the projection and `@` is the current value.

The `expression` within the brackets specifies the initial value of `$`, if `expression` is omitted the initial value of `$` is the first value of the array/object and the first iteration has `@` starting at the second value.

The resulting value of the reducing-projection is the accumulated value (or the value of the last iteration).

examples:
```
search([1, 2, 3, 4], [%1].$+@) -> 11
search([1, 2, 3, 4], [%].$+@) -> 10
search([5, 1, 2, 4], [%].(($<@) && $) || @) -> 1
search([[a, 1], [b, 2], [c, 3]], [% {} ].merge($, {@[0] : @[1]})) -> {a: 1, b: 2, c: 3}
search([true, false, true], [%].$ && @) -> false
search([true, false, true], [%].$ || @) -> true
search([1,2,2,2,3,4], [%{}].merge($, {@ : $[@] + 1})) -> {1: 1, 2: 3, 3: 1, 4: 1}
```

**Reply by springcomp** [reply]
Maybe `%` is not a good choice on my part. I can easily see this used as the `modulo` operator instead.

**Reply by springcomp** [reply]
Or it could serve both purposes because the context is unambiguous.

**Comment by springcomp** [comment]
Based on [this comment](https://github.com/jmespath-community/jmespath.spec/discussions/82#discussioncomment-3110187) I think the reduce operation is _not_ a projection. In that it produces a single value and does not expect downstream expressions to apply on multiple items from the result of the operation.

**Reply by innovate-invent** [reply]
I think there is a misunderstanding, reduce does expect downstream expressions to apply. For `[% init ].expr`, expr is applied to each value in the current context.

**Reply by springcomp** [reply]
I see now.

What you are referring to as the _reduce projection_ is only the `bracket-specifier` `[% init]` which does not actually reduce anything until downstream expressions would get applied against each projected value to accumulate the result.

I was referring to the _reduce operation_ as the whole process yielding a single result value after applying the reduce logic.

So, yes, you are right. I agree the _reduce_ `bracket-specifier` would indeed be a projection.

**Comment by springcomp** [comment]
For consistency reason, since we have `map()` function, I would tend to favor a corresponding `reduce()` function rather than implement dedicated syntax for this operation.

**Reply by innovate-invent** [reply]
I still think `map()` is a kludge to deal with the mistake of the mapping projection filtering nulls.

**Reply by springcomp** [reply]
I have started a series of proof of concepts to explore implementing the reduce feature with various proposals. The idea would be to actually find out the pros and cons and limits to each proposal.

## `reduce()` function and `lambda-expression`

I have actually fleshed out a [reasonably complete implementation](https://github.com/springcomp/JmesPath.Net/tree/feature/reduce-lambda-expression) of a [`reduce()`](https://github.com/springcomp/JmesPath.Net/blob/feature/reduce-lambda-expression/src/jmespath.net/Functions/ReduceFunction.cs) function using [`lambda-expression`](https://github.com/springcomp/JmesPath.Net/blob/79caed18c9a5d9573473c6952e71541e8e4c26d1/src/jmespath.net.parser/JmesPathParser.y#L158-L192) for the actual reduction sequence. 

The result is a neat and easy implementation, with some caveats.

- any: `` search( reduce(@, `false`, (acc, cur) => acc || cur), [true, false, false]) -> true ``
- any: `` search( reduce(@, `false`, (acc, cur) => acc || cur), [false, false, false]) -> false ``
- all: `` search( reduce(@, `true`, (acc, cur) => acc && cur), [true, false, false]) -> false ``
- all: `` search( reduce(@, `true`, (acc, cur) => acc && cur), [true, true, true]) -> true ``

### Pros

The work done for implementing the `let()` function is paying off as we can reuse the notion of _scopes_ that has been introduced and is specifically used in function implementation. Here a new scope is created to map the named arguments of the `lambda-expression` to their actual values for the accumulated and current values in each iteration of the evaluated input array.

### Cons

The grammar to implement `lambda-expression` is slightly cumbersome as it is actually ambiguous 🤔.
I managed to limit the conflicting syntax to a _shift/reduce_ conflict that fortunately does the right thing by virtue of the parser doing a _shift_ by default in that case.

The `JmesPathLambdaExpression` AST node is only a placeholder for named values and a delayed expression.

### Misceallaneous

In my implementation, I made `JmesPathLambdaExpression` an implicit `expression-type`. Maybe only the expression after the `=>` token should be and `expression-type` and maybe we should let the user specify this explicitly.

The current implementation does not supply any context to as arguments to the lambda expression, except for the `{ acc: acc, cur: cur}` scope. As a result, the value for the `@` `current-node` is undefined and effectively `null`. This is not intuitive to users. Maybe `@` should be the same value as `cur` in the `lambda-expression`?

## `[% expression]` syntax-based `reduce-expression`

I have started an effort to implement a pure syntax-based [`reduce-expression`](https://github.com/springcomp/JmesPath.Net/tree/feature/reduce-syntax) syntax,
in order to  contrast the previous implementation.

This would support syntaxes such as:

- any: `` search( [% `false`] -> $ || @, [true, false, false]) -> true ``
- any: `` search( [% `false`] -> $ || @, [false, false, false]) -> false ``
- all: `` search( [% `true`] -> $ || @, [true, false, false]) -> false ``
- all: `` search( [% `true`] -> $ || @, [true, true, true]) -> true ``

### Pros

The syntax is more terse than using `lambda_expression`.
It uses a [`reduce-expression`](https://github.com/springcomp/JmesPath.Net/blob/e540ec44d4a84527267f4e308a06efb95b32a622/src/jmespath.net.parser/JmesPathParser.y#L355-L370) rule in the grammar. A `reduce-expression` cannot easily be a projection (see Cons below). It uses two new tokens: `[%` `T_REDUCE` and `->` `T_ARROW`.

A third token `$` `T_DOLLAR` supports the `root-node` expression that evaluates to the currently accumulated value in a reduce operation.

The `root-node` expression does nothing but extract the current _accumulated_ value from a stack of accumulators, This stack is maintained by the `reduce-expression` instances in the JMESPath expression. I used a stack as I expect `reduce-expression` instances to be nested in complex expressions.

The stack of accumulators is a new structure that needs to be attached to:

- `reduce-expression` AST node.
- `root-node` AST node.

This is easily done using the Visitor pattern.

### Cons

The first hurdle to overcome for the grammar is that a `sub-expression` does not allow the syntaxes that we currently expect to have on the right hand side of a reduce projection. So I tried using a new `->` arrow token which makes a nice syntax that look like a lambda overall. The side effect is that the bracket portion does not have any meaning on its own in the AST and the overall expression is not a projection after all.

To make it more consistent with existing syntax, I also elected to add a new alternative to `index-expression`:

```abnf
index-expression =/ expression reduce-expression
```
That enables such syntax as:

- all: `` search( foo[%] -> $||@, [true, false, false]) -> true ``

However, the grammar itself is not very consistent as a `reduce-expression` is not a `bracket-specifier`. I guess this is a minor detail, but I often find a design should reflect its internal beauty or elegance in the implementation, which, currently, feels a bit botched together.

### Miscellaneous

In my current implementation I have ignored the dual purpose of the `$` `root-node` token. I’m not sure we should have a token that changes its behaviour based upon where it occurs in a JMESPath expression.

In my current implementation, the stack of accumulators also implement setting and retrieving the value of the _current_ accumulator (_i.e_ the accumulator currently sitting at the top of the stack). A cleaner implementation would separate both concerns.

## `reduce_projection`

This alternate solution consists in having two different syntaxic constructs instead of a single `reduce_expression` as above. The goal is to support expressions like:


- sum: `` search( [% `0`].$ + @, [1, 2, 3]) -> 6 ``
- any: `` search( [% `false`]. ($ || @), [true, false, false]) -> true ``
- any: `` search( [% `false`]. ($ || @), [false, false, false]) -> false ``
- all: `` search( [% `true`].  ($ || @), [true, false, false]) -> false ``
- all: `` search( [% `true`]. ($ || @), [true, true, true]) -> true ``

Notice that you need `paren-expression` arround `or-` and `and-expression` because the `.` token has higher precedence than the `||`  OR operator. We believe this is a mistake in the grammar but let’s try to stay backwards compatible. The first expression is currently not supported but could be very easily using [`arithmetic_expression`](https://github.com/jmespath-community/jmespath.spec/discussions/25#discussioncomment-3277652) for which I created a [sample implementation](https://github.com/springcomp/JmesPath.Net/blob/d331bffddd6b10858eb001e9ad25f3cfbf63180c/src/jmespath.net.parser/JmesPathParser.y#L121-L125).

### Pros

This syntax is closer to the original envisioned proposal where the bracket-part would be a _projection_ that projects its values to the right-hand-side of the `sub-expression`.

This is easy to re-introduce support into the grammar of `paren-expression` without regressions.

```abnf
expression =/ reduce-projection

sub-expression =/ expression "." arithmetic-expression
sub-expression =/ expression "." paren-expression

reduce-projection = "[%" expression "]"
```
### Cons

My current .NET port however is blocked short after this because a _projection_ is a concept that carries a list of values, each of which applies to the downstream expressions. Up until now, this was sufficient because the projected values all go through downstream expressions and the result itself is a projection.

The reduce operation is special in that the result of the projection is a single value after reduction. So to get this applied to a projection object, the downstream expressions – or objects in the AST node – would need to know that a reduction is needed, to carry the requested state holding the `acc` and `cur` values.

Looking at the [Python implementation](https://github.com/jmespath/jmespath.py/blob/bbe7300c60056f52413603cf3e2bcd0b6afeda3d/jmespath/visitor.py#L272-L294) we can see that _projections_ do not carry any state. They just project theirs values to the downstream expressions and collect the result.

To support reduction, refactoring is needed.

I will try and implement this and update this post accordingly.

**Reply by innovate-invent** [reply]
This is excellent! Thanks for this :)

I am not clear why `sub-expression` is a subset of `expression`. Do you know where the ambiguity is with the missing components?

**Reply by springcomp** [reply]
Well I think I get a feel for it. Let's focus on what would happen if the syntax were:

```abnf
sub-expression = expression "." expression
```

First, some non-starters:

Clearly , `and-`, `or-`,  `not-` and `comparator-expression` _predicates_ do not make a lot of sense to me as the RHS expressions. They are intended to be used in comparisons and discard the LHS expression of the `sub-expression`. This rules out:

```abnf
expression "." and-expression
expression "." or-expression
expression "." not-expression
expression "." comparator-expression
```

Subscription using the `bracket-specifier` array syntax also should be ruled out in my opinion as the syntax is not intuitive and already exists without the `.` dot token in the `index-expression` rule. So that eliminates further:

```abnf
expression "." bracket-specifier
```
One obvious no-op redundant syntax using `current-node` should also be ruled out as well, if only for the fact that it is useless. So out goes:

```abnf
expression "." current-node
```

Two of the remaining last candidates would discard their LHS expression if they were allowed. This is, I suspect, the reason why those are ruled out as well:

```abnf
expression "." raw-string
expression "." literal
```
Most of the remaining ones are already represented by the `sub-expression` rule. The terminal precedence rules always conspire to make _their_ LHS collapse into the left part of the `sub-expression`. Thus, it makes sense to prevent ambiguity in the grammar by forbidding them:

```abnf
expression "." sub-expression
expression "." pipe-expression
```
This leaves us with **one** remaining candidate, which although not harmful, is actually useless, given the terminal precedence rules. Again, I think that is why it was excluded.

```abnf
expression "." paren-expression
```

If we were to revisit the grammar in a non compatible way, I think I would not decide much differently. Maybe only allowing the following syntaxes, even though they provide no definite value.

```abnf
expression "." paren-expression
expression "." literal
expression "." raw-string
```

But I would not fight on a hill to get them supported in the syntax 😏.

**Reply by innovate-invent** [reply]
I think I am imagining a different operator precedence, but the spec doesn't actually specify the precedence of some of the operators.
I was imagining something like this in increasing binding strength:
parenthesis: )
pipe: |
sub-expression: .
or: ||
and: &&
comparator: > < == != <= >=
unary not: !
rbracket: ]

`foo.bar || baz` would be equivalent to `foo.(bar || baz)` rather than `(foo.bar) || baz` but it appears that is not how it is implemented in the javascript library. This is annoying as I would expect to be able to write

`search(a.b.c.d.e.f.g.h.i || j, { "a": {"b": {"c": {"d": {"e": {"f": {"g": {"h": {"i": false, "j":2 }}}}}}}}}) -> 2`

rather than having to write

`search(a.b.c.d.e.f.g.h.i || a.b.c.d.e.f.g.h.j, { "a": {"b": {"c": {"d": {"e": {"f": {"g": {"h": {"i": false, "j":2 }}}}}}}}}) -> 2`

we currently can't even write 

`search(a.b.c.d.e.f.g.h.(i || j), { "a": {"b": {"c": {"d": {"e": {"f": {"g": {"h": {"i": false, "j":2 }}}}}}}}}) -> 2`

the closest we can get is the pipe operator

`search(a.b.c.d.e.f.g.h | i || j, { "a": {"b": {"c": {"d": {"e": {"f": {"g": {"h": {"i": false, "j":2 }}}}}}}}}) -> 2`

but then you have to consider more complex examples like

`search([*].a.b.c.d.e.f.g.h | i || j, [{ "a": {"b": {"c": {"d": {"e": {"f": {"g": {"h": {"i": false, "j":2 }}}}}}}}},{ "a": {"b": {"c": {"d": {"e": {"f": {"g": {"h": {"i": false, "j":2 }}}}}}}}}]) -> [2,2]` 

which I don't readily see a way to make work

I don't know what the benefit of having `.` being a higher binding strength than other operators is.

**Reply by springcomp** [reply]
Somes miscellaneous contrasting examples:

Return the highest length from a set of arrays, containing the number `2`:

- **syntax**: `` search ( [? contains(@, `2`)]|[% `0`] -> max([$, length(@)]), [[1, 2, 3], [2, 3], [1, 4]] ) -> 3 ``
- **lambda**: `` search ( reduce([?contains(@, `2`)], `0`, (acc, cur) => max([acc, length(cur)])), [[1, 2, 3], [2, 3], [1, 4]] ) -> 3 ``

**Reply by innovate-invent** [reply]
> This leaves us with one remaining candidate, which although not harmful, is actually useless, given the terminal precedence rules. Again, I think that is why it was excluded.
> `expression "." paren-expression`

I am not sure I fully understand this. If we add `paren-expression` to `sub-expression`, that would solve the issue I mention above. We could additionally add `arithmetic-expression`, which would cover 90% of the expected reduce operations anyway.

This would allow `[%].$+@` and `[%].($ || @)` ('sum' and 'any' operations respectively)

**Reply by springcomp** [reply]
I understand, that is a good idea.

**Reply by innovate-invent** [reply]
if arethmetic operators have a stronger binding than `.` then it should work. This would be different than the or/and operators.
This is all coming from a limited understanding of how this is being parsed, hopefully I am not way off.

I think you ninja edited something and I am responding to nothing :yum:

**Reply by springcomp** [reply]
# TL;DR

## `reduce()` function and `lambda-expression`

A `reduce()` function is straightforward to implement as it is reasonably scoped. The work done while implementing the `let()` function is paying dividends as we can reuse the concept of _scopes_ out of the box.

The current implementation uses a `lamba-expression` to eliminate some magic around naming the `acc` and `cur` values but this is not strictly necessary. If we were to introduce the `root-node` expression instead, a new stack of accumulated values would need to be introduced, so that the accumulator could be set by the `reduce()` function and accessed by the downstream `root-node` expression.

## `[% expression] -> expression` syntax-based `reduce-expression`

A `reduce-expression` that encompasses the bracket portion and the reducing logic is possibly syntactically. We need to introduce an additional token as the whole expression cannot be a `sub-expression`.

Implementing reduce as syntax requires that a new stack of accumulated values be passed on to each interested expression, In the current implementation, only the `reduce-expression` and corresponding `root-node` expression would need to access to the current value of the accumulator.

## `reduce-projection`

A `reduce-projection` is only possible if the reducing mechanism is built into the core concept of projection in the implementation. This is needed for a projection usually does not carry any state, so downstream expressions have no way to understand that in this special case, reduction needs to happen.

A `reduce-projection` has the syntax we initially envisioned, but needs to be special-cased.

**Comment by springcomp** [comment]
For the record, here is an alternate proposal for a `reduce()` function.

```
any reduce(array[any], object $seed, expression->expression->any $expr)
```

The `$seed` object would be pushed into the current scope (much like the first argument of the `let()` function does).

The first property of the `$seed` object would be the seed to the reduce operation.

In a reduce operation, the accumulated value starts as the seed. So the name of the first property in the `$seed` object would be the name under which the accumulated value is known and could be referred to in the `$expr` expression.

Thus, `any()` and `sum()` could be emulated by the following expressions, assuming an appropriate input JSON array.

- any: `` reduce(@, { acc: `false` }, &acc || @) `` (is there any _truthy_ value in the input array)
- sum: `` reduce(@, { acc: `0` }, &acc + @) `` (subject to [`arithmetic-expression`](https://github.com/jmespath-community/jmespath.spec/discussions/25#discussioncomment-3277652))

## Pros

Reuses concepts already introduced by the `let()` function.

## Cons

None 🤣

**Reply by springcomp** [reply]
Interesting to note that this proposal [was considered](https://github.com/jmespath/jmespath.php/pull/30#issuecomment-264668696)  early on in the design of JMESPath.

**Comment by springcomp** [comment]
Some of us are keen to try and support a reduce operation through pure syntax instead of relying on a function.
The current syntax being proposed looks like this:

- `` [% `false`].$ || @ ``

Here `$` would refer to _the accumulated value within a reduce operation_ and is irrelevant to this present comment. Let’s focus instead on the overall syntax. This syntax uses a `sub-expression` whose rhs is an `or-operation`. That is not currently not supported in the grammar.

Instead of thinking of the reduce being split into a left _projection_ that projects its values to a right _reduce operation_, let’s invert the process and have a `reduce-expression` accept an array and operate on it.

- ``  @.[% `false`, &$ || @] ``

Here the bracket part is not a `multi-select-list`. Instead, it is a `reduce-expression`that is defined like so:

```abnf
expression =/ reduce-expression
sub-expression =/ expression "." reduce-expression

reduce-expression = "[%" [ expression ] "," expression-type "]"
```

Popular functions, like `any()` or `sum()` could be emulated using the following syntax:

- any `` [% `false`, &$ || @]  ``
- sum `` [% `0`, &$ + @]`` (subject to [`arithmetic-expression`](https://github.com/jmespath-community/jmespath.spec/discussions/25#discussioncomment-3277652) being part of the language)

**Reply by springcomp** [reply]
Maybe overloading bracket notation on the rhs is confusing.

**Reply by springcomp** [reply]
`expression-type` is currently only supported for function arguments. However, the grammar shown above still works perfectly fine and has not ambiguity.

**Comment by springcomp** [comment]
In light of a [newly proposed `let-expression`](https://github.com/jmespath/jmespath.jep/pull/18) it might be worth revisiting those proposals here, and include a feature that centers around:

- Defining anonymous functions witha arguments.

Maybe a new `reduce()` function that would be invoked like this:

```
reduce(
  array,
  `0`, 
  ($acc, $cur) => & $cur + $acc
)
```

Where `` $acc => $acc + @ `` would be a `lambda-expression` defining an anonymous function with a two arguments named `$acc` and `$cur` whose implementation is an `expression-type` that sums those two values.

`reduce()` itself would take three arguments, 
- An array,
- An initial value for the accumulator
- The `lambda-expression`

And it would call the `lamba-expression` for all items in the array, gathering the result as an input to the next iteration.

**Reply by springcomp** [reply]
Here is an [initial attempt](https://github.com/jmespath/jmespath.jep/issues/25) at this initiating feedback on this proposal.
